cmd/process-agent/main_windows.go

```diff
@@ -8,117 +8,62 @@
+	"github.com/DataDog/datadog-agent/pkg/util/winutil/messagestrings"
+	"github.com/DataDog/datadog-agent/pkg/util/winutil/servicemain"
 
 	"golang.org/x/sys/windows/svc"
-	"golang.org/x/sys/windows/svc/debug"
-	"golang.org/x/sys/windows/svc/eventlog"
-	"golang.org/x/sys/windows/svc/mgr"
 )
 
-var elog debug.Log
-
 const (
 	// ServiceName is the service name used for the process-agent
 	ServiceName  = "datadog-process-agent"
 	useWinParams = true
 )
 
-type myservice struct {
+type service struct {
 	globalParams *command.GlobalParams
 }
 
-func (m *myservice) Execute(args []string, r <-chan svc.ChangeRequest, changes chan<- svc.Status) (ssec bool, errno uint32) {
-	const cmdsAccepted = svc.AcceptStop | svc.AcceptShutdown
-	changes <- svc.Status{State: svc.StartPending}
-	changes <- svc.Status{State: svc.Running, Accepts: cmdsAccepted}
-
-	exit := make(chan struct{})
-
-	go func() {
-		for {
-			select {
-			case c := <-r:
-				switch c.Cmd {
-				case svc.Interrogate:
-					changes <- c.CurrentStatus
-					// Testing deadlock from https://code.google.com/p/winsvc/issues/detail?id=4
-					time.Sleep(100 * time.Millisecond)
-					changes <- c.CurrentStatus
-				case svc.Stop, svc.PreShutdown, svc.Shutdown:
-					elog.Info(0x40000006, ServiceName)
-					changes <- svc.Status{State: svc.StopPending}
-					///// FIXME:  Need a way to indicate to rest of service to shut
-					////  down
-					close(exit)
-					break
-				default:
-					elog.Warning(0xc000000A, fmt.Sprint(c.Cmd))
-				}
-			}
-		}
-	}()
-	elog.Info(0x40000003, ServiceName)
-
-	// On Windows, the SCM will require that dependent services stop. This means that when running
-	// `Restart-Service datadogagent`, windows will try to stop the Process Agent, and then to be helpful it will immediately start it again.
-	// However, if Process Agent is not configured to be running it will exit immediately, which `Restart-Service` will report as an error.
-	// To avoid the error on a successful exit we ensure that we are in the RUNNING state long enough for `Restart-Service` or other tools to consider
-	// the restart successful.
-	exitGate := time.After(5 * time.Second)
-	defer func() { <-exitGate }()
-
-	runAgent(m.globalParams, exit)
-
-	changes <- svc.Status{State: svc.Stopped}
-	return
+func (s *service) Name() string {
+	return ServiceName
 }
 
-func runService(globalParams *command.GlobalParams, isDebug bool) {
-	var err error
-	if isDebug {
-		elog = debug.New(ServiceName)
-	} else {
-		elog, err = eventlog.Open(ServiceName)
-		if err != nil {
-			return
-		}
-	}
-	defer elog.Close()
-
-	run := svc.Run
-	if isDebug {
-		run = debug.Run
-	}
-	elog.Info(0x40000007, ServiceName)
-
-	service := &myservice{
-		globalParams: globalParams,
-	}
+func (s *service) Init() error {
+	// Nothing to do, kept empty for compatibility with previous implementation.
+	return nil
+}
 
-	err = run(ServiceName, service)
+func (s *service) Run(ctx context.Context) error {
```

Comment:
Would it be possible to add some tests for the windows service behavior? It would be nice to have some code coverage in this area